src/users/index.ts

```diff
@@ -18,35 +46,151 @@ export interface Users {
    */
   getAssignedRoles: (userId: string) => Promise<Record<string, Role>>;
+
+  db: DBUsers;
+  oidc: OIDCUsers;
+}
+
+/** Operations supported for namespaced 'db' users.*/
+export interface DBUsers extends UsersBase {
   /**
-   * Assign roles to a user.
+   * Retrieve the roles assigned to a user.
    *
-   * @param {string | string[]} roleNames The name or names of the roles to assign.
-   * @param {string} userId The ID of the user to assign the roles to.
-   * @returns {Promise<void>} A promise that resolves when the roles are assigned.
+   * @param {string} userId The ID of the user to retrieve the assigned roles for.
+   * @returns {Promise<Record<string, Role>>} A map of role names to their respective roles.
    */
-  assignRoles: (roleNames: string | string[], userId: string) => Promise<void>;
+  getAssignedRoles: (userId: string, opts?: GetAssingedRolesOptions) => Promise<Record<string, Role>>;
+
+  create: (userId: string) => Promise<string>;
+  delete: (userId: string) => Promise<boolean>;
+  rotateKey: (userId: string) => Promise<string>;
+  activate: (userId: string) => Promise<boolean>;
+  deactivate: (userId: string) => Promise<boolean>;
+  byName: (userId: string) => Promise<UserDB>;
+  listAll: () => Promise<UserDB[]>;
+}
+
+type GetAssingedRolesOptions = {
+  includePermissions?: boolean;
+};
+
+/** Operations supported for namespaced 'oidc' users.*/
+export interface OIDCUsers extends UsersBase {
   /**
-   * Revoke roles from a user.
+   * Retrieve the roles assigned to a user.
    *
-   * @param {string | string[]} roleNames The name or names of the roles to revoke.
-   * @param {string} userId The ID of the user to revoke the roles from.
-   * @returns {Promise<void>} A promise that resolves when the roles are revoked.
+   * @param {string} userId The ID of the user to retrieve the assigned roles for.
+   * @returns {Promise<Record<string, Role>>} A map of role names to their respective roles.
    */
-  revokeRoles: (roleNames: string | string[], userId: string) => Promise<void>;
+  getAssignedRoles: (userId: string, opts?: GetAssingedRolesOptions) => Promise<Record<string, Role>>;
 }
 
 const users = (connection: ConnectionREST): Users => {
+  const base = baseUsers(connection);
+
   return {
     getMyUser: () => connection.get<WeaviateUser>('/users/own-info').then(Map.user),
     getAssignedRoles: (userId: string) =>
       connection.get<WeaviateRole[]>(`/authz/users/${userId}/roles`).then(Map.roles),
+    assignRoles: (roleNames: string | string[], userId: string) => base.assignRoles(roleNames, userId),
+    revokeRoles: (roleNames: string | string[], userId: string) => base.revokeRoles(roleNames, userId),
+    db: db(connection),
+    oidc: oidc(connection),
+  };
+};
+
+const db = (connection: ConnectionREST): DBUsers => {
+  const ns = namespacedUsers(connection);
+
+  const allowCode = (code: number): ((reason: any) => boolean) => {
```

Comment:
I like this method but the `reason: any` here is a concern type-safety-wise 🤔

I see that this is used in `.catch` methods so can we change it to perform a type guard inside `allowCode` for when the error thrown is a known unexpected status code? E.g.
```typescript
const allowCode = (code: number): ((reason: any) => boolean) => (reason) => {
  if (reason instanceof WeaviateUnexpectedStatusCodeError) {
    return reason.code === code;
  };
  throw reason;
}
```
This way, if the thrown error is something else (somehow) then that error is still propagated up to the user appropriately. wdyt?

Reply:
Makes sense to me 👍